### PR TITLE
[SPARK-39961][SQL] DS V2 push-down translate Cast if the cast is safe

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, DS V2 push-down translate `Cast` only if the ansi mode is true.
In fact, if the cast is safe(e.g. cast number to string, cast int to long), we can translate it too.

This PR will call `Cast.canUpCast` so as we can translate `Cast` to V2 `Cast` safely.

Note: The rule `SimplifyCasts` optimize some safe cast, e.g. cast int to long, so we may not see the `Cast`. 

### Why are the changes needed?
Add the range for DS V2 push down `Cast`.

### Does this PR introduce _any_ user-facing change?
'Yes'.
`Cast` could be pushed down to data source in more cases.

### How was this patch tested?
Test cases updated.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @huaxingao cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala [hunk 1]
```diff
@@ -88,7 +88,8 @@ class V2ExpressionBuilder(e: Expression, isPredicate: Boolean = false) {
       } else {
         None
       }
-    case Cast(child, dataType, _, true) =>
+    case Cast(child, dataType, _, ansiEnabled)
+      if Cast.canUpCast(child.dataType, dataType) || ansiEnabled =>
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:92]: nit: 4-space indentation?

**Review by huaxingao — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:92]: +1 for @huaxingao 's comment. 

In addition, let's switch the predicate order, @beliefer .
```scala
- if Cast.canUpCast(child.dataType, dataType) || ansiEnabled =>
+ if ansiEnabled || Cast.canUpCast(child.dataType, dataType) =>
```

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:92]: Got it.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master for Apache Spark 3.4.

**Comment by beliefer** [comment]
@dongjoon-hyun @huaxingao @cloud-fan Thank you!

**Pull request closed without merging**